src/main/scala/com/fulcrumgenomics/alignment/NeedlemanWunschAligner.scala

```diff
@@ -147,13 +173,40 @@ class NeedlemanWunschAligner(val scoringFunction: (Byte,Byte) => Int,
     * @return an [[Alignment]] object representing the alignment
     */
   protected def generateAlignment(query: Array[Byte], target: Array[Byte], scoring: Matrix[Int], trace: Matrix[Direction]): Alignment = {
-    var i = query.length
-    var j = target.length
     var currOperator: CigarOperator = null
     var currLength: Int = 0
     val elems = new mutable.ArrayBuffer[CigarElem]()
 
-    while (i != 0 || j != 0) {
+    // Always start at the end of the query
+    var i = query.length
+
+    // For the target sequence, start at the end for Global, or the highest scoring cell in the last row for Glocal
+    var j = this.mode match {
+      case Global =>
+        target.length
+      case Glocal =>
+        var (maxScore, maxIndex) = (Int.MinValue, -1)
+        forloop(from=0, until=target.length) { i =>
+          val score = scoring(query.length-1, i)
+          if (score > maxScore) {
+            maxScore = score
+            maxIndex = i
+          }
+        }
```

Comment:
How's this for an ugly alternative:
`Stream.range(start=0, end=target.length).zipWithIndex.maxBy(tup => scoring(query.length, tup._1))._2`

If you really want to make it fancy, you can make this into an impicit method (`indexOfMax`) on a `TraversableOnce`.

Or you can keep it readable and as is.

Reply:
Lol.  Cryptic and likely several orders of magnitude slower due to all the intermediate object creation.